Delete domain fixes

millerdev opened the pull request:
Fixes for errors encountered while deleting domains.

NOTE changes behavior of `hard_delete_forms_and_cases_in_domain` to **hard-delete ALL forms and cases** in the domain, as reflected in the prompt presented by the command. Previously that command only deleted forms and cases that had previously been soft-deleted.

More details in https://dimagi-dev.atlassian.net/browse/SAAS-13147

:blowfish: Review by commit.

## Safety Assurance

### Safety story

All changes in this PR should only affect code related to permanently deleting domains, forms, and cases, which is only done as a one-off management command.

### Automated test coverage

~No new tests.~ Adapted tests around `delete_domain` and `hard_delete_forms_and_cases_in_domain` to what those commands actually do (and say they do).

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

If this PR is rolled back the migration should be reverted on all environments where it has been applied so as not to conflict with future migrations created in the `formplayer_api` app.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

millerdev committed:
Do not crash on missing `demo_restore_id`

Prevents error: AttributeError: 'dict' object has no attribute 'demo_restore_id'

millerdev committed:
MigratedReminder before AutomaticUpdateRule

Prevents error: django.db.models.deletion.ProtectedError: ("Cannot delete some instances of model 'AutomaticUpdateRule' because they are referenced through a protected foreign key: 'MigratedReminder.rule'", <QuerySet [<MigratedReminder: MigratedReminder object (...)>, <MigratedReminder: MigratedReminder object (...)>]>)

millerdev committed:
Improve hard_delete_forms_and_cases_in_domain

Remove code duplication and use more efficient deletion routines.

millerdev committed:
Add migration to delete unused formplayer tables

dannyroberts commented:
```
======================================================================
  FAIL: corehq.apps.domain.tests.test_delete_domain:TestHardDeleteSQLFormsAndCases.test_hard_delete_cases_none_to_delete
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/mnt/commcare-hq-ro/corehq/apps/domain/tests/test_delete_domain.py", line 1099, in test_hard_delete_cases_none_to_delete
      self.assertEqual(len(CommCareCase.objects.get_case_ids_in_domain(self.domain2.name)), 1)
  AssertionError: 0 != 1
```

millerdev committed:
Add test for delete_domain command (with cases)

millerdev committed:
Update hard_delete_forms_and_cases_in_domain test

Contrary to its messaging, previously that command only hard-deleted forms and cases that had previously been soft-deleted. It now hard-deletes all forms and cases in the domain.

millerdev committed:
Fix hard delete forms test

Same theory as previous commit
